Installation - Support "activate first" w/setup UI

totten opened the pull request:
Overview
----------

This adds support for installing CiviCRM-Backdrop through the [Setup API](https://docs.civicrm.org/dev/en/latest/framework/setup/new-installer/). Most notable changes:

* Better layout and more useful options on the installation screen
* Extensible installation logic - under the hood, this installer can initialize settings, toggle extensions, etc.
* Reusable installation logic - can run the same installation code for CLI and web UI
* No need to bypass regular module activation or bypass router (no need for `install/index.php` entry-point)

Before
------

The sysadmin downloads the CiviCRM module - but they must *not* install it. Instead, navigate to this path:

```
http://example.org/modules/civicrm/install/index.php?civicrm_install_type=backdrop
```

This is a special, standalone entry-point.

After
-----

After downloading the CiviCRM module, you simply enable it like any module. Then use any link to open a CiviCRM page (e.g. there are links in the post-install status-message and in the main menu). The page will display the setup screen.

![Screen Shot 2020-07-03 at 11 17 37 PM](https://user-images.githubusercontent.com/1336047/86506346-72a4f480-bd83-11ea-88c8-3c986e4a7ba6.png)

Slightly better demo: http://think.hm/tmp/Backdrop-Civicrm-Setup.mp4

Technical Details
-----------------

1. The old install screen is still available as a fallback (in case there's some edge-case that got missed).
2. The basic gist of this change is:
    * Ensure that all CiviCRM-Backdrop hooks fail gracefully (in the absence of `civicrm.settings.php`). Most already do - there were just a couple oddballs.
    * Update the handling of `civicrm/*` URLs. If CiviCRM cannot initialize, then show the [civicrm-setup web installer](https://docs.civicrm.org/dev/en/latest/framework/setup/new-installer/#web-installer-api).
3. This is the same API used in [the newer Civi-WP installer](https://github.com/civicrm/civicrm-wordpress/blob/5.28/civicrm.php#L1189-L1207). (*Note: For historical reasons, the Civi-WP has a few extra contingencies that aren't needed here.*)
4. When testing/`r-run`ning, here are some of things that I tried:
    * Enable the module and try to view it as regular/non-admin user. Regular users cannot perform installation.
    * Install via web UI 
    * Install via CLI (`cv core:install -vv --cms-base-url=http://bcmaster.bknix:8001`), with the `civicrm.module` enabled or disabled
    * Install in US English or French.
4. During development, my workflow went a bit like this:
    1. Create Backdrop site with CiviCRM codebase -- but without installing. (*I specifically used a modified `civibuild` with the `backdrop-clean` type - ie hack `app/config/backdrop-clean/install.sh` to skip the scripted `civicrm_install`, etal. Then run `civibuild create bcmaster` or `civibuild reinstall bcmaster`*)
    2. Login.
    3. Make a new DB snapshot (eg `civibuild snapshot bcmaster`). (This way my login-session is part of the snapshot.)
    4. Repeatedly do the following:
        * Edit code
        * Use the installer - and see how it goes.
        * Cleanup: `./bin/cv.phar core:uninstall -f ; civibuild restore bcmaster ; rm -rf ~/bknix/build/bcmaster/web/files/civicrm/`
5. This depends on another PR for civicrm-core: https://github.com/civicrm/civicrm-core/pull/17749

totten committed:
All `civicrm.module` hooks should fail gracefully on uninitialized system

totten committed:
Installation - Use "Setup" API with "activate-first" pageflow

herbdool commented:
I tested and left feedback here https://github.com/civicrm/civicrm-core/pull/17749#issuecomment-654960606

totten committed:
Installation - Move setup reminder. Avoid redundant messages.

Before
------

Every time `civicrm_initialize()` is called with a failure, it enqueues another status message
which prompts the admin to setup CiviCRM.

This leads to some quirks - e.g. if you activate `civicrm.module` in the admin GUI, and then use
the web-based installer, and then finally return to a Backdrop page - you may see some
old/stale notifications prompting you to setup CiviCRM (which is weird... since you just did that).

After
-----

The setup message is displayed in two ways:

1. In `hook_enable`, it displays the message one time.
2. In `hook_requirements`, it displays the message if initialization fails.

totten committed:
All `civicrm.module` hooks should fail gracefully on uninitialized system
